adaptive/utils.py

```diff
@@ -98,3 +99,43 @@ def __call__(self, *args, **kwargs):
                     msg = f"The attribute '{name}' should be of type {type_}, not {type(x)}."
                     raise TypeError(msg)
         return obj
+
+
+def _default_parameters(
+    function, function_prefix: str = "function.", start_index: int = 1
+):
+    sig = inspect.signature(function)
+    defaults = {
+        f"{function_prefix}{k}": v.default
+        for i, (k, v) in enumerate(sig.parameters.items())
+        if v.default != inspect._empty and i >= start_index
+    }
+    return defaults
+
+
+def assign_defaults(
+    function, df, function_prefix: str = "function.", start_index: int = 1
+):
+    defaults = _default_parameters(function, function_prefix, start_index)
+    for k, v in defaults.items():
+        df[k] = len(df) * [v]
+
+
+def partial_function_from_dataframe(function, df, function_prefix: str = "function."):
+    if function_prefix == "":
+        raise ValueError(
+            "The function_prefix cannot be an empty string because"
+            " it is used to distinguish between function and learner parameters."
+        )
+    kwargs = {}
+    for col in df.columns:
+        if col.startswith(function_prefix):
+            k = col.split(function_prefix, 1)[1]
+            vs = df[col]
+            v, *rest = vs.unique()
+            if rest:
+                raise ValueError(f"The column '{col}' can only have one value.")
```

Comment:
Just realized that this means that any default values attached to the learner's function will be repeated in the dataframe for every point the learner evaluates!

I guess the reason for this is to facilitate combining results from learners where the default values are different, but this clause shows that `partial_function_from_dataframe` will fail in that case.

Is it possible to attach "metadata" to an entire dataframe (xarrays have this) to avoid the redundancy?

Reply:
Using constant columns is a standard practice.

It is correct that using the `partial_function_from_dataframe` directly won't work and I would also expect it to not work.

The `BalancingLearner` does a `df.groupby` operation and then uses `partial_function_from_dataframe`.

Since pandas 1.0 there is a metadata attribute `df.attrs` however, they won't combine easily. I think one of the biggest advantages to this new "learners to dataframes" feature is precisely being able to combine learners easily, so I prefer the current implementation.

Reply:
We can use the `category` datatype:
<img width="981" alt="image" src="https://user-images.githubusercontent.com/6897215/190242735-c739ccec-84e3-4508-9587-d451389689f6.png">

Reply:
Not saving parameters that the learner doesn't vary is another option to consider. It makes the code simpler and avoids handling special cases.

Reply:
I think we should keep it in, there aren't that many special cases and the code is quite simple.